sql/src/main/java/io/crate/execution/dsl/phases/HashJoinPhase.java

```diff
@@ -0,0 +1,147 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.dsl.phases;
+
+import io.crate.exceptions.UnsupportedFeatureException;
+import io.crate.execution.dsl.projection.Projection;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.Symbols;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+
+public class HashJoinPhase extends JoinPhase {
+
+    private final List<Symbol> leftJoinConditionInputs;
+    private final List<Symbol> rightJoinConditionInputs;
+
+    private final Collection<DataType> leftOutputTypes;
+    private final int leftSize;
+
+    public HashJoinPhase(UUID jobId,
+                         int executionNodeId,
+                         String name,
+                         List<Projection> projections,
+                         @Nullable MergePhase leftMergePhase,
+                         @Nullable MergePhase rightMergePhase,
+                         int numLeftOutputs,
+                         int numRightOutputs,
+                         Collection<String> executionNodes,
+                         @Nullable Symbol joinCondition,
+                         List<Symbol> leftJoinConditionInputs,
+                         List<Symbol> rightJoinConditionInputs,
+                         Collection<DataType> leftOutputTypes,
+                         long leftSize) {
+        super(
+            jobId,
+            executionNodeId,
+            name,
+            projections,
+            leftMergePhase,
+            rightMergePhase,
+            numLeftOutputs,
+            numRightOutputs,
+            executionNodes,
+            JoinType.INNER,
+            joinCondition);
+        assert joinCondition != null : "JoinCondition for HashJoin cannot be null";
+        this.leftJoinConditionInputs = leftJoinConditionInputs;
+        this.rightJoinConditionInputs = rightJoinConditionInputs;
+        this.leftOutputTypes = leftOutputTypes;
+        if (leftSize > Integer.MAX_VALUE) {
+            throw new UnsupportedFeatureException("number of rows for the smaller table must be <= max integer value");
+        }
+        this.leftSize = (int) leftSize;
+    }
+
+    public HashJoinPhase(StreamInput in) throws IOException {
+        super(in);
+
+        leftJoinConditionInputs = Symbols.listFromStream(in);
+        rightJoinConditionInputs = Symbols.listFromStream(in);
+
+        int numLeftOutputTypes = in.readVInt();
+        if (numLeftOutputTypes > 0) {
```

Comment:
Maybe it would make sense to add a  `DataTypes.listFromStream` as well? - Doesn't have to be in this PR, or even on this branch. Just a suggestion.